### PR TITLE
Add onPause\onResume persistence to searchView. Fixes issue #3627

### PR DESCRIPTION
**Pull request by AntoineGS**
Still learning the ropes with Android\Kotlin coming from Delphi so if something isn't Kotlin-y enough let me know :)

### CODE REVIEW

**Review by arkon — COMMENTED** [review]
Considering a lot of the logic is actually duplicate, maybe we could put it in the parent class (`NucleusController`)? I think there's actually a few other places that could take advantage of  this too that aren't called out in the original issue.

**Comment by AntoineGS** [comment]
Good point, I was making a conservative commit as my first contribution so I was hesitant to start refactoring a dozen files (I think there are 15 files with a very similar onCreateOptionsMenu that inherit from NucleusController).
I'll refactor and push another commit for this.

**Comment by arkon** [comment]
Makes sense. If you want to keep the PR small, you could add the base logic and only apply it one controller, then do another PR (or more) to apply it to others.

**Comment by AntoineGS** [comment]
Ok so this is small-ish commit.
I sub-classed BaseController so that I can work on the conversion without breaking the build or creating a massive PR.
Once everything is converted I can merge SearchableNucleusController into BaseController and make the necessary declaration changes.
SearchView has a lot of nitpicky behaviors, like initializing its query to "" after you have set a value and thus losing the set value.

OK so while writing this I realized I might be able to get around this whole loading\loaded issue by posting the query restore to the queue which will hopefully mean it will execute after the initialization to "".. If this works it should simplify the class.

**Comment by AntoineGS** [comment]
Ok so it does not really remove the need to know when the SearchView is loaded, when resuming the activity I would still need to know or the nonCommitedQuery will be set to "".
Plus it makes the UI jitter as it gets set to "" and the to the actual search.
Between the two the committed implementation works better, let me know if you have any comments. 
It is unfortunately pretty complex for something that I had figured would be easy to implement, though making it centralized makes it easier to maintain.

**Comment by AntoineGS** [comment]
Hey @arkon, once this is merged I will do the other controllers (in case you were waiting on more commits in this PR before merging as I am not planning to). 
Probably you are just busy with other stuff, if so no worries ;)

**Comment by arkon** [comment]
I'll be sure to review this PR this weekend.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/BaseController.kt [hunk 1]
```diff
@@ -121,7 +121,7 @@ abstract class BaseController<VB : ViewBinding>(bundle: Bundle? = null) :
      * [expandActionViewFromInteraction] should be set to true in [onOptionsItemSelected] when the expandable item is selected
      * This method should be called as part of [MenuItem.OnActionExpandListener.onMenuItemActionExpand]
      */
-    fun invalidateMenuOnExpand(): Boolean {
+    open fun invalidateMenuOnExpand(): Boolean {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/BaseController.kt:124]: It actually probably makes sense to move all of this search stuff from this class to your new `SearchableNucleusController` class, as it's not actually relevant for controllers without a `SearchView`.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    private fun acceptEmptyQuery(): Boolean {
+        return when (currentSearchViewState) {
+            SearchViewState.COLLAPSING, SearchViewState.FOCUSED -> true
+            else -> false
+        }
+    }
+
+    private fun setCurrentSearchViewState(newState: SearchViewState, fromState: SearchViewState? = null) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:155]: Nitpick: let's use `to` and `from` (which we use elsewhere too, like in `syncActivityViewWithController`).

File: app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceController.kt [hunk 3]
```diff
@@ -3,13 +3,7 @@ package eu.kanade.tachiyomi.ui.browse.source
...
 import android.app.Dialog
 import android.os.Bundle
-import android.view.LayoutInflater
-import android.view.Menu
-import android.view.MenuInflater
-import android.view.MenuItem
-import android.view.View
-import android.view.ViewGroup
-import androidx.appcompat.widget.SearchView
+import android.view.*
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/browse/source/SourceController.kt:6]: You can turn this off from Preferences > Editor > Code Style > Kotlin > Imports

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    protected open fun onSearchMenuItemActionExpand(item: MenuItem?) {
+    }
+
+    protected open fun onSearchMenuItemActionCollapse(item: MenuItem?) {
+    }
+
+    /**
+     * During the conversion to SearchableNucleusController (after which I plan to merge its code
+     * into BaseController) this addresses an issue where the searchView.onTextFocus event is not
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:188]: I sort of wonder if we _should_ merge them. I guess we could, but I feel like it's making `BaseController` a lot more complicated than it needs to be too.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * Store the query text that has not been submitted to reassign it after an onResume, UI-only
+     */
+    protected var nonSubmittedQuery: String = ""
+
+    /**
+     * To be called by classes that extend this subclass in onCreateOptionsMenu
+     */
+    protected fun commonCreateOptionsMenu(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:42]: Nitpick: maybe just name this `createOptionsMenu`?

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        menu: Menu,
+        inflater: MenuInflater,
+        menuId: Int,
+        searchItemId: Int,
+        queryHint: String = "",
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:47]: Let's actually make this a `@StringRes` `Int` to force uses of it to be localized strings.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            }
+
+            if (restoreCurrentQuery) {
+                // Restoring a query the user had submitted
+                if (query.isNotBlank()) {
+                    searchItem.expandActionView()
+                    searchView.setQuery(query, true)
+                    searchView.clearFocus()
+                    onSearchViewQueryTextChange(query)
+                    onSearchViewQueryTextSubmit(query)
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:101]: Seems a bit odd to fire both.

**Review by arkon — COMMENTED** [review]
Overall looks pretty good, just some minor comments.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/BaseController.kt (hunk 1 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/BaseController.kt:124]: Good point, I've taken note of it to remove it once all views with SearchView are moved onver to SearchableNucleusController.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 5 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:42]: Done

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 6 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:47]: I learned something new, thanks ;)

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 7 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:101]: The idea behind it is that some implementations (ie local searches) are triggered by the `onSearchViewQueryTextChange` event and do not need to implement the submit event.
Firing only the `onSearchViewQueryTextSubmit` event would require more implementations.
I do not mind doing it, let me know if you prefer it with only the submit being fired.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 2 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:155]: Done!

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 4 above)
Comment (AntoineGS) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:188]: Makes sense, there is no real need to merge them if there are view using the 'vanilla' `BaseController`. Plus it keeps all the complicated state logic in its own unit so it is more obvious that is it for the search capabilities only.

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt (hunk 7 above)
Comment (arkon) [reply app/src/main/java/eu/kanade/tachiyomi/ui/base/controller/SearchableNucleusController.kt:97]: Probably fine for now. We can address it later if it seems problematic.

**Comment by arkon** [comment]
Thanks again! We can follow up on this work in separate PRs.

**Pull request merged**